lib/resqued/listener_pool.rb

```diff
@@ -0,0 +1,92 @@
+require "resqued/listener_proxy"
+require "resqued/listener_state"
+
+module Resqued
+  class ListenerPool
+    include Enumerable
+
+    # Public: Initialize a new pool, and store state in the given master's state.
+    def initialize(master_state)
+      @master_state = master_state
+      @listener_proxies = {}
+      # If this master is replacing an old one, there will be listeners in the state already.
+      @master_state.listener_states.each do |pid, ls|
+        @listener_proxies[pid] = ListenerProxy.new(ls)
+      end
+    end
+
+    # Public: Iterate through all active ListenerProxy instances.
+    def each(&block)
+      @listener_proxies.values.each(&block)
+    end
+
+    # Public: Number of active listeners.
+    def size
+      @listener_proxies.size
+    end
+
+    # Public: Initialize a new listener, run it, and record it as the current listener. Returns its ListenerProxy.
+    def start!
+      listener_state = ListenerState.new
+      listener_state.options = {
+        :config_paths => @master_state.config_paths,
+        :old_workers => map { |l| l.running_workers }.flatten,
+        :listener_id => next_listener_id,
+      }
+      listener = ListenerProxy.new(listener_state)
+      listener.run
+      @master_state.listener_states[listener.pid] = listener_state
+      @listener_proxies[listener.pid] = listener
+      @master_state.current_listener_pid = listener.pid
+      return listener
+    end
+
+    # Public: Remove the given pid from the set of known listeners, and return its ListenerProxy.
+    def delete(pid)
+      @master_state.listener_states.delete(pid)
+      return @listener_proxies.delete(pid)
```

Comment:
I think the explicit `return` can be dropped here?

Reply:
I'm going to leave it in, I want it to be obvious that the result of this `delete()` call is intended to be returned.